### ISSUE TITLE
Explore lazy property merge

### ISSUE THREAD

**Issue by Chalarangelo** [Chalarangelo/jsiqle#8]
### Background

`Function.prototype.length` can be an indicator if a function expects one argument or more.  If we can reliably use it, we could merge `lazyProperties` with regular ones and force jsiqle to evaluate the `length` of each one deciding how to handle it. 

Argument lengths are pretty logical, with the exception of `0`. In that case it's ambiguous due to it referring to either 0 arguments or variadic arguments. Practically, restricting the API to not use fully variadic arguments could be a solution.

### Proposal

Such a change will mean merging the lazy properties API into the regular properties API. It will also mean altering the lazy property definition API from this:

```js
{
  lazyProperty: ({ models, serializers }) => value => value.x;
}
```

to this:

```js
{
  lazyProperty: (value, { models, serializers }) => value.x;
}
```

This will also require some argument binding in such a way that the first argument is bound to the schema's output.

Distinguishing between properties (lazy, non-lazy) is easy using `Function.prototype.length`:

- An argument `length` of **0** would be handled as a regular (non-lazy) property (e.g. static properties).
- An argument `length` of **1** would be handled as a regular (non-lazy) property. 
- An argument `length` of **2** would be handled as a lazy property.
- Other argument `length` values would throw an error.

### Considerations

Method mapping, while in the same ballpark is not as easy to perform, due to methods expecting any number of arguments.

On the same note, lazy methods will remain in their special key indefinitely due to them being impossible to map reasonably. Checking for single argument length and a return type of function is the only idea so far and it won't work due to methods being able to take any number of arguments and return any type, even functions.

Due to the above note, the `lazyProperties` key should not be deprecated entirely but throw a warning in itself that it's not the best practice any more. This will ensure backwards compatibility and make the API less confusing for beginners.

**Comment by Chalarangelo** [comment]
## Update

Implementation has been pushed to [v2](https://github.com/Chalarangelo/jsiqle/milestone/1). As such, this will be a breaking change and will completely deprecate the old API. It's not worth supporting both in the next major version.

**Comment by Chalarangelo** [comment]
Resolved in #22.
